### PR TITLE
[data] Fix and re-enable EliminateBuildOutputBlocks rule

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

There was a bug in the `EliminateBuildOutputBlocks` rule that caused a perf regression for `iter_batches_benchmark_single_node` and other benchmarks. For read op with additional splits, which has `BuildOutputBlocksMapTransformFn -> BlockMapTransformFn`, we cannot drop the `BuildOutputBlocksMapTransformFn`. This PR fixes this issue by combing the 2 steps in one MapTransformFn.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/planner/plan_read_op.py [hunk 1]
```diff
@@ -55,23 +56,34 @@ def _splitrange(n, k):
...
-    """
-    assert additional_output_splits > 1
-    for block in blocks:
-        block = BlockAccessor.for_block(block)
-        offset = 0
-        split_sizes = _splitrange(block.num_rows(), additional_output_splits)
-        for size in split_sizes:
-            yield block.slice(offset, offset + size, copy=True)
-            offset += size
+class BuildOutputBlocksWithAdditionalSplit(MapTransformFn):
```
Comment (c21) [inline python/ray/data/_internal/planner/plan_read_op.py:59]: This one looks bit awkward right now, let's make splitblock an operator later.

**Review by c21 — APPROVED** [review]

File: python/ray/data/_internal/planner/plan_read_op.py (hunk 1 above)
Comment (raulchen) [reply python/ray/data/_internal/planner/plan_read_op.py:59]: yeah, also it's also weird that previously BuildOutputBlock was not the last transform fn and caused the bug.

**Comment by raulchen** [comment]
Hey @zhe-thoughts this PR is ready for merge. The CI has passed before, and I only pushed a small comment update after that.

**Comment by zhe-thoughts** [comment]
@raulchen Master is un-frozen for now. Please merge as usual, and I will approve the cherry-pick PR

**Pull request merged**